### PR TITLE
r.geomorphon: implement a new "one-off" mode

### PR DESCRIPTION
**Pull request by infrastation**
The commit explains the proposed change in detail, please find some higher-level comments below.

The new r.geomorphon feature has been stable for a few weeks. It looks good to me, although somebody else might see issues in it. In any case, it is ready for a review.

Most importantly, it works and it has a use case: the machine-readable output of r.geomorphon becomes input to [Geomorphon Profiler](https://gitlab.com/geomorphon-hunters/GeomorphonProfiler) (another piece of work done in the course of this project), which provides data and illustrations for a research project of a PhD student.

Most of the new code in `profile.c` is generic enough to be useful in other plugins. I did not go as far as making it a part of the library, but if you wish to do it later, it should not be difficult.

### CODE REVIEW

File: raster/r.geomorphon/main.c
Comment (wenzeslaus) [inline raster/r.geomorphon/main.c:473]: Just a note that I appreciate this form of check esp. that it is reported only when the region is large. It should fit well with usual workflows in GRASS GIS.

**Review by wenzeslaus — CHANGES_REQUESTED** [review]
This is a very interesting addition. Seeing it in action at the sprint and in #1096 helped me to see its value. Thank you.

I would be interested if other people think that this sets a good precedent for other modules, new or old, to allow for new applications and a different level of debugging (as shown in #1096).

I did only a review of the inputs now. So, more reviews are still needed.

File: raster/r.geomorphon/main.c
Comment (nilason) [inline raster/r.geomorphon/main.c:65]: duplicated `const`

(This is the only complain by clang on r.geomorphon module, with this PR.)

File: raster/r.geomorphon/main.c
Comment (infrastation) [reply raster/r.geomorphon/main.c]: done

File: raster/r.geomorphon/main.c
Comment (infrastation) [reply raster/r.geomorphon/main.c:475]: The use case magnitude is gigacells, so when the region size overshoots it takes many minutes just to run through the buffer dry.

File: raster/r.geomorphon/main.c
Comment (infrastation) [reply raster/r.geomorphon/main.c]: removed

File: raster/r.geomorphon/main.c
Comment (infrastation) [reply raster/r.geomorphon/main.c:466]: `calc_pattern()` always looks into the centre of a region cell. As far as I understand it, in this case the best approach is to align the region cells with the raster cells to avoid resampling, but the alignment logic in `g.region` does not seem to be available as a library function. In order to achieve the effect without duplicating the code, the first version of Geomorphon Profiler just calls `g.region` before `r.geomorphon`. This could be a valid improvement for the next version.

File: raster/r.geomorphon/main.c
Comment (infrastation) [reply raster/r.geomorphon/main.c:282]: Fine, removal of `-1` is in my working copy and will be added to the pull request soon.

File: raster/r.geomorphon/main.c
Comment (infrastation) [reply raster/r.geomorphon/main.c:349]: Let me take some time to see how that solution space matches the problem space.

**Comment by infrastation** [comment]
There is one other point not raised elsewhere. The proposed code uses `G_fopen_new()` to open the profile data file for writing. One problem with that is it overwrites existing files even if `--overwrite` is not given. Some time ago I thought this behaviour depended on the current directory so it refused to overwrite an existing file in _some_ cases. However, the only way it reproduces now is that it overwrites existing files in _all_ cases.

Based on other usage examples of `G_fopen_new()` I suspect it might not be the right function for this job, or maybe this code does not use it properly. For example, r.out.png uses just `fopen()` and maybe that's what this code should do too. This would also solve the second problem, in that it would be possible to give an absolute filename and to save the profile data outside of the GRASS data directory. Currently Geomorphon Profiler works around this by redirecting stdout of r.geomorphon into a temporary file. It could look better, but it works.

If you think this specific point must be improved before merging, please explain how, otherwise I would fine to leave it for the next version.

**Comment by wenzeslaus** [comment]
Right, `G_fopen_new` is not the right function. That for things inside the location/mapset structure. Perhaps the documentation is misguiding.

You just want the plain `fopen` (some modules seem to use `freopen` if the output can be stdout, I can't quickly figure out how that fits together). So, yes, r.out.png is probably right. v.db.select is another example. The command line parser in GRASS checks options which are created using `G_OPT_F_OUTPUT` for presence of that file and does not proceed if the file exists unless `--overwrite` is given (as you would expect).

File: raster/r.geomorphon/main.c
Comment (infrastation) [reply raster/r.geomorphon/main.c]: This is done now: `-1` no longer exists, `coordinates=` requests the one-off mode, `profiledata=` is optional as before.

**Comment by infrastation** [comment]
Regarding the parser dependencies, the problem space seems to be as follows (REQ means "required", OPT means "optional"):
```text
* REQ: "elevation"
* REQ: [1] exactly one of:
	* OPT: (not multires), implicit, zero or more of:
		* OPT: "forms"
		* OPT: "ternary"
		* OPT: "positive"
		* OPT: "negative"
		* OPT: "intensity"
		* OPT: "exposition"
		* OPT: "range"
		* OPT: "variance"
		* OPT: "elongation"
		* OPT: "azimuth"
		* OPT: "extend"
		* OPT: "width"
	* OPT: (multires) "prefix", enables:
		* OPT: [2] "step"
		* OPT: [2] "start"
	* OPT: (one-off) "coordinates", enables:
		* OPT: [2] "profiledata", enables:
			* OPT: [2] "profileformat"
* OPT: "search"
* OPT: "skip"
* OPT: "flat"
* OPT: "dist"
* OPT: "comparison"
* OPT: "-m"
* OPT: "-e"
```
The current implementation (with improvised checks) more or less covers it, less proper warnings about parameters that are no-op because their dependency is not satisfied (see [2] above).

Having looked through the parser dependency function list, I understand that [1] can be implemented through `G_option_required()` with a long list of hard-coded arguments. I am not sure how many `G_option_excludes()` calls it will take to make the three main modes mutually exclusive. The four parameters that hinge on other parameters will require one call to `G_option_requires()` each, which seems to be the easiest part. That said, working mode selection would still need to use variables to remain simple.

Does this look correct? Do you expect some or all of these changes to be a part of this feature?

**Review by wenzeslaus — CHANGES_REQUESTED** [review]
I tested more the outputs and I think some changes are needed. It is lot of points, but hopefully overall not so much.

1. The default text output (the ASCII picture) is nice, but it needs to be documented. Possibly even in the output itself, I'm not sure about that. Of course the manual page is always fine.
2. The default text output should go to stdout. User asked for the info, so stdout is appropriate. Which brings me to...
3. The default text output should be really one of the output formats, perhaps named `text` or `plain`. (We don't have a good precedence for the name. I have the same issue in a current  v.db.select PR. The point is that it is human readable plain text. So maybe `text` since it is text?) `profileformat` would either have a default `text` or default to it when not provided.
4. Not providing `profiledata` and `profiledata=-` should be the same.
5. I get no search results for `format_ver_major`, while `format_version_major` seems to be used quite a bit. Perhaps if you take care to introduce version numbers into your format, you may as well spell it out, too.
6. Any reason for some of the keys having both float and string values? Sorry, I don't see an immediate advantage. Is it for a latitude-longitude CRS so that DMS format is in the string?
7. How to reproduce the ASCII graphics with the other outputs (by looking at the outputs, not code). Perhaps that would be basic thing to document.
8. `elevname` -> `name` or `elevation_name`

I actually started to review the formats to check whether the XML is one of these attribute-heavy ones, but it is not, so +1 for that (respect). :+1:

**Comment by wenzeslaus** [comment]
> ... Does this look correct?

If you put it like this, then perhaps the current state is good enough.

> Do you expect some or all of these changes to be a part of this feature?

I was really just thinking about the pieces you are adding.

The handling of many options in r.geomorphon was written before the functions for option dependencies were added and the functions don't cover all cases, so it is really just that you should take advantage of these where you can (for easier maintenance, easier translations, ...), but it is not that you need to express everything using those.

Bottom line: I though you don't know about these functions. Now you do and can see where it makes sense to use them. I think that you can use them at least in between the new options.

> ...I am not sure how many `G_option_excludes()` calls it will take to make the three main modes mutually exclusive...

Is `G_option_exclusive()` what you are looking for?

**Comment by infrastation** [comment]
Thank you for the review. The new profile-specific options now use parser dependencies and the plugin rejects some invalid combinations at an early stage, for example:

```text
ERROR: Option <profiledata> requires <coordinates>
ERROR: Option <profileformat> requires <coordinates>
ERROR: Option <coordinates> is mutually exclusive with all of prefix
ERROR: Option <coordinates> is mutually exclusive with all of forms
ERROR: Option <coordinates> is mutually exclusive with all of extend
```

(In the latter case you might want to add angle braces into the message to make it more consistent.)

As to the other points, please see below.

1, 2, 3 and 7: The message with the final landform and the ASCII grid was intended for the interactive session as a means to run a quick sanity check without overwhelming the scroll buffer or redirecting the output. It would be wrong to deliver it as one of the machine-readable formats. I thought it would be a convenient little addition to have, but from the feedback it is clear it interferes with the main use case too much, so I have moved these messages to Geomorphon Profiler for now.

4: Done, although in future this might come back to bite when someone adds a binary output format or the text encoding is not good for the user's terminal encoding.

5: Done.

6: The reason for the twin string+float format was indeed to enable visualization in LatLong a while ago. Since then the visualization code had evolved not to depend on this, so the proposed serialization code now produces only the string format. Can be made float-only.

8: Done.

**Comment by wenzeslaus** [comment]
> Thank you for the review. The new profile-specific options now use parser dependencies and the plugin rejects some invalid combinations at an early stage, for example:

I need to run it, but the code looks good.

> 1, 2, 3 and 7: The message with the final landform and the ASCII grid was intended for the interactive session as a means to run a quick sanity check without overwhelming the scroll buffer or redirecting the output. It would be wrong to deliver it as one of the machine-readable formats. I thought it would be a convenient little addition to have, but from the feedback it is clear it interferes with the main use case too much, so I have moved these messages to Geomorphon Profiler for now.

Well, I actually liked the output, just thought it should be handled as another format. Some machine readable, some not. That's fine. See e.g. r.info versus r.info -g. The modules should give nice outputs to interactive command line user if possible/applicable. Machine non-readable as default is good too. User should specify explicitly for the desired format.

> 4: Done, although in future this might come back to bite when someone adds a binary output format or the text encoding is not good for the user's terminal encoding.

True, but I don't think neither likely now. It will be also an usage issue consistent across several modules, so it would be more or less expected.

> 6: The reason for the twin string+float format was indeed to enable visualization in LatLong a while ago. Since then the visualization code had evolved not to depend on this, so the proposed serialization code now produces only the string format. Can be made float-only.

They are numbers, so why not all float? When machine reading degrees, floating point is expected and normal (see, e.g., GeoJSON).

Thanks for the updates. I still need to get back to you on the other parts. Just to be clear, @nilason or anybody else, feel free to jump in even with small comments. This PR has many different aspects.

**Comment by nilason** [comment]
Clang is happy and so am I – for any maintenance and improvement of this module, to which I not infrequently return to in my line of work. I don't have the expertise to dissect the algorithms, but I didn't experience any issues trying out this PR.

**Comment by infrastation** [comment]
The output uses float format now. I have made a look at r.info, let me think about it. There are other, much more technically demanding aspects of the proposed change, if these require amendments, it would be best to figure it out before long.

**Comment by infrastation** [comment]
After some consideration it becomes clear that `r.info` is not a good model: it prints human-readable messages by default and it prints machine-readable data in two incompatible formats depending on which of the `-h/-g/-s/-r/-e` flags were given. If you try to redo `r.info` parser declarations to follow the same `data=format` style of options as is implemented for `g.geomorphon` here, you should see it does not work well in the first place. If you still think the proposed implementation must have a better arrangement, please provide a better example.

I believe this fully addresses all review points made so far, and this pull request is ready for a review or for merging to make it into GRASS 8.0, as it was originally planned.

**Comment by wenzeslaus** [comment]
> ...If you try to redo `r.info` parser declarations to follow the same `data=format` style of options as is implemented for `g.geomorphon` here, you should see it does not work well in the first place.

Okay, so although no longer related to this PR, but if you don't mind, can you please elaborate on why do you think it would not work for `r.info`? Or is that what you meant? I want to do this for `v.db.select` as proposed in #1121, but `format=value` is actually something we want to move to in general, so if you spotted any flaws I'm eager to hear them (although maybe not happy to hear them :-).

> ...please provide a better example.

I would say `v.db.select` as proposed in #1121, but I think I'm missing what you consider bad in this concept. Anyway, no action need here.

> I believe this fully addresses all review points made so far, and this pull request is ready for a review or for merging to make it into GRASS 8.0, as it was originally planned.

You are right. I was hoping someone else would test it too, but it is hard to fully evaluate without the subsequent visualization. I wish I could fully review the core code from theoretical perspective, but your analyses and visualizations are convincing enough.

I make this a final call for reviews and otherwise I will re-check this and merge it soon.

**Review by wenzeslaus — APPROVED** [review]

**Comment by infrastation** [comment]
Thank you for the review. Let me explain the comment about `r.info` above. Having one option with multiple values instead of multiple flags is indeed better when the multiple flags are mutually exclusive. I agree on that. The matter is, the current set of `r.info` flags would not map to such an option well because the flags are not mutually exclusive and every permutation would effectively require a new value:
```shell
r.info info=all-freeform                 # the current default
r.info info=history-freeform             # -h
r.info info=raster-shell                 # -g
r.info info=range-shell                  # -r
r.info info=stats-shell                  # -s
r.info info=extended-shell               # -e
r.info info=raster-range-shell           # -gr
r.info info=raster-stats-shell           # -gs
r.info info=raster-extended-shell        # -ge
r.info info=range-stats-shell            # -rs
r.info info=range-extended-shell         # -re
r.info info=stats-extended-shell         # -se
r.info info=raster-range-stats-shell     # -grs
r.info info=raster-range-extended-shell  # -gre
r.info info=raster-stats-extended-shell  # -gse
r.info info=range-stats-extended-shell   # -rse
r.info info=all-shell                    # -grse
```
Essentially this includes a mapping from an O(n) bitmap space into an O(2**n) name space. Anyway, whether you see it the same way or differently, this indeed is out of scope of this pull request.

I have just tested that the proposed changes work well after rebasing on the current master branch.

**Comment by wenzeslaus** [comment]
> The matter is, the current set of `r.info` flags would not map to such an option well because the flags are not mutually exclusive and every permutation would effectively require a new value: [...]

Thanks for analyzing that. That would be indeed a mess, but maybe I was not clear in what I meant initially. I don't think one option needs to handle this. For, r.info it seems that there are at least two options, one for format, another for content. The one for format (`format=freeform|shell`) could be implemented independently from the `-hrse` which could be additive also as an option (e.g., `content=stats,extended` or some combinations which are currently not possible) and applied only with format != freeform. However, the question is if all these are even needed, if getting that info does not cost much, you may as well output all to formats such as JSON.

Back to r.geomorphon, there my point was that there is always the same information exported with the notable exception of the human readble output (your ASCII graphics), so no separate `content=` option would be needed and it is really just about the format. Yes, the human readable output is really different from the rest, but it includes what is best for the human eye and leaves out the rest (the others could even include the ASCII graphics besides the numbers), so I would say it contains the same information represented in the best possible way in the given format (you could have a format char which outputs just one letter signifying the landform - perhaps not your algorithm profiling info, but still one-off mode).

Perhaps r.info was not a good example and v.db.select in #1121 would be a better one. I picked r.info because, unlike v.db.select or r.univar, it has a very different human readable output from the machine readable one in the two cases `r.info abc` and `r.info -g abc` (I didn't really mean to involve all the specialties of the current, messy r.info interface). 

> I have just tested that the proposed changes work well after rebasing on the current master branch.

Thanks. I didn't mean for this to sit here for so long after my review. Sorry about that.

**Comment by infrastation** [comment]
Thank you very much!

**Comment by wenzeslaus** [comment]
Hm, there is actually a more important problem here with the command line parsing. The basic test now fails because of a confusion in the command line parameters.

History (in r.info) and pygrass Module fill in the defaults (so `profiledata="-" profileformat="json" is put there`). However, when such command is executed, it fails with:

```
ERROR: Option <profiledata> requires <coordinates>
ERROR: Option <profileformat> requires <coordinates>
```

I assume this happened only after I suggested to use the automated check declaration instead of a manual checks.

The solution is either going back to the manual check or not providing defaults which may actually make sense in this case.

On an unrelated note, I have realized that profileformat should be profile_format etc.

**Comment by infrastation** [comment]
Thank you for pointing this out, I confirm this effect reproduces in my environment. I didn't realize how to run the test before.

If the issue was `profiledata` only, it would be straightforward just to lose the default (as it was before) without thinking about it too much. But `profileformat` made me suspect that it is wrong to fill in all defaults explicitly in the first place.

If you look at the current `r.geomorphon` running through the GUI in the traditional (raster in raster out) mode, it works fine. This is because the GUI displays default values for both the above two options and `comparison`, but when it comes to the actual command-line arguments, these options do not appear there unless the user selects a non-default value.

I could not find a good example quickly, but consider an imaginary plugin with the following options:
```text
r.out.image
    format=<png|jpeg> (mandatory, no default)
    pngcompression=<1~9> (optional, default=9)
    jpegquality=<20~100> (optional, default=85)
    jpegprogressive=<true|false> (optional, default=false)
```
With proper arrangement of parser dependencies (it is an error to specify `pngcompression` for JPEG and so on) this plugin would never be able to pass a test that always specifies all defaults explicitly, regardless if `format` has a default and which value you set it to. There is a subtle distinction between _having_ a default value for an option and having the option _specified_ on the command line. As far as I can see it now, the current parser specifies and implements this distinction well, but the code that runs this test does not. Does it make sense?

In terms of practical next steps:
* If snake case would be more appropriate for the option names, please go ahead and make it snake case.
* Regarding the defaults, I have tried to find quickly where the explicit defaults come from, but got lost in `class Module`, so I cannot propose a solution right away. If you agree that the explicit default values are wrong and can remove them quickly without opening a can of worms in other plugins/tests, that would make the most sense to me.
* If you agree, but cannot, it should be fine to remove the two default values in `r.geomorphon` with a detailed comment that this is actually a workaround for a problem in `class TestCase` (or some such), and to raise the problem in a more appropriate place (the mailing list?).
* Let me know in case you need more input.

**Comment by infrastation** [comment]
What do you think would be the best way to fix this?

**Comment by wenzeslaus** [comment]
Created #2029 so this "optional with default" issue is not forgotten and also to suggest that a more broader solution might be needed.

**Pull request merged**
